add Event Hubs samples

joshgav opened the pull request:
Adds samples for managing Event Hubs namespaces and hubs, as well as using event hubs to send and receive events.

@devigned can you please review? Also, test of send/receive doesn't work at the moment, will look at main repo for ideas, do you have suggestions?

fixes #159

devigned commented:
@joshgav what is the error you are receiving? Doesn't work is kind of abstract.

joshgav commented:
Good point :). Receive doesn't actually get a message in the test run, wondering if I've chosen the starting offset correctly, and if I need to have it wait.

devigned commented:
Receive spins up a listener and continues on. You will have to wait to receive messages. Take a look at 
https://github.com/Azure/azure-event-hubs-go/blob/master/hub_test.go#L324-L355. Which provides an example of send and receive with wait.

devigned commented:
@joshgav here are the start of stream end of stream consts: https://github.com/Azure/azure-amqp-common-go/blob/master/persist/checkpoint.go#L29-L37

If you don't supply a starting offset, `Receive` will assume you mean start of stream.

joshgav commented:
Thanks @devigned! Adjusted to use the constant and tests now pass :)

joshgav commented:
Hi @devigned and @marstr, this is now complete with Send, Receive, and Receive with EPH examples, all passing tests. Can you please take a look? Would like to merge as soon as possible so we can reference from docs, preferably before EOD 6/29. Thank you in advance!

devigned commented on eventhubs/receive_eph.go:
```diff
@@ -0,0 +1,114 @@
+	}
+
+	// register the handler with the Event Processor
+	close, err := p.Receive(handler)
```
@joshgav I liked your suggestion of changing this to register and unregister handler. I'm going to make that change to better illustrate the difference between Receive and Start*

joshgav replied:
:+1: Great! Should I wait on that change to merge? If it will be a little while we can update this when its ready.

devigned commented on eventhubs/send_events.go:
```diff
@@ -0,0 +1,49 @@
+	// 	hub.Send(ctx, eventhubs.NewEventFromString(text))
+	// }
+
+	// send message to hub
```
perhaps, point out this is sending round robin

marstr commented on eventhubs/receive_eph.go:
```diff
@@ -0,0 +1,114 @@
+
+	// set up wait group to wait for expected message
+	var wg sync.WaitGroup
+	wg.Add(1) // wait for one message
```
I'd like to steer away from using a `sync.WaitGroup` here, because they error out if they get decremented more times than expected.

If it were me, I'd use channels, something like this:

``` Go
receivedEvent := make(chan struct{})

handle := func(ctx context.Context, event *eventhubs.Event) error {
    fmt.Printf("received: %s\n", string(event.Data))
    receivedEvent <- struct{}{}
    return nil
}

// stuff

<-receivedEvent
```

Both approaches are prone to leaking the resources captured via the closure of this func. However, using channels will prevent misused code from panicing. There's probably some argument to be had here, because health checks on paniced code may be easier to detect than memory leaks, but for the sake of example code I'd like them to be a little more robust to people tinkering with the number of messages sent, etc.

marstr commented on eventhubs/receive_eph.go:
```diff
@@ -0,0 +1,114 @@
+	}
+
+	// finally, start the Event Processor with a timeout
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
I'd get rid of this redeclaration of `ctx`, and just have the timeout encapsulate the registration process as well. See the comment at the first declaration of `ctx` for more description of what I'm talking about. :)

marstr replied:
At very least, this one should derive from the original ctx instead of being derived from `context.Background()` again.

marstr commented on eventhubs/receive_eph.go:
```diff
@@ -0,0 +1,114 @@
+	}
+
+	// don't exit till a message has been processed
+	wg.Wait()
```
One of the benefits of switch to using channels instead of wait group would mean you could do this:

``` Go
switch {
case <-receivedEvent:
    // Intentionally Left Blank
case err := <-ctx.Done():
    log.Fatal("Event not received quickly enough:", err )
}
```

marstr commented on eventhubs/receive_eph.go:
```diff
@@ -0,0 +1,114 @@
+
+	// imports within this repo
+	"github.com/Azure-Samples/azure-sdk-for-go-samples/helpers"
+	storage "github.com/Azure-Samples/azure-sdk-for-go-samples/storage"
```
[nit] For import paths where the last component is the same as what you wanted, I wouldn't create an alias. To me it actually gets in the way of identifying the conflict between `storage` in the samples repository and `storage` in the event hubs repository.

marstr commented on eventhubs/receive_eph.go:
```diff
@@ -0,0 +1,112 @@
+
+	// don't exit till a message has been processed
+	wg.Wait()
+	p.Close(context.Background())
```
I'm confused by this line. @devigned, why does `p.Close` take a context? If it's for correlative purposes, doesn't using context.Background() eliminate its usefulness?

joshgav commented:
I addressed all your comments @marstr and made the changes you requested. Can you PTAL again? Thanks!

marstr commented on eventhubs/eventhubs_test.go:
```diff
@@ -0,0 +1,82 @@
+	log.Printf("Receive()\n")
+	Receive()
+	log.Printf("ReceiveViaEPH()\n")
+	ReceiveViaEPH()
```
Instead of creating a new context in `ReceiveViaEPH`, I'd pass in the one created here. :)

joshgav replied:
Fair enough, working on this now, thank you for helping adopt better practices in the samples!

joshgav commented:
@marstr I moved context to top scope and fixed up a couple other items, PTAL.

I'm looking into the Travis setup now, the failure isn't cause of this I don't think.
